tornado/web.py

```diff
@@ -997,7 +997,8 @@ def finish(self, chunk=None):
             if self._status_code in (204, 304):
                 assert not self._write_buffer, "Cannot send body with %s" % self._status_code
                 self._clear_headers_for_304()
-            elif "Content-Length" not in self._headers:
+            elif ("Content-Length" not in self._headers and
```

Comment:
No reason.  I just didn't read this clause carefully enough to notice it.  Done.
(AFAICT, `self._clear_headers_for_304` should apply here as well).